{"expand":"renderedFields,names,schema,transitions,operations,editmeta,changelog","id":"12332120","self":"https://issues.apache.org/jira/rest/api/latest/issue/12332120","key":"DERBY-1219","fields":{"issuetype":{"self":"https://issues.apache.org/jira/rest/api/2/issuetype/1","id":"1","description":"A problem which impairs or prevents the functions of the product.","iconUrl":"https://issues.apache.org/jira/images/icons/issuetypes/bug.png","name":"Bug","subtask":false},"timespent":null,"project":{"self":"https://issues.apache.org/jira/rest/api/2/project/10594","id":"10594","key":"DERBY","name":"Derby","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/projectavatar?pid=10594&avatarId=10122","24x24":"https://issues.apache.org/jira/secure/projectavatar?size=small&pid=10594&avatarId=10122","16x16":"https://issues.apache.org/jira/secure/projectavatar?size=xsmall&pid=10594&avatarId=10122","32x32":"https://issues.apache.org/jira/secure/projectavatar?size=medium&pid=10594&avatarId=10122"},"projectCategory":{"self":"https://issues.apache.org/jira/rest/api/2/projectCategory/10090","id":"10090","description":"DB related projects","name":"DB"}},"fixVersions":[{"self":"https://issues.apache.org/jira/rest/api/2/version/11187","id":"11187","description":"","name":"10.2.1.6","archived":false,"released":true,"releaseDate":"2006-10-02"}],"aggregatetimespent":null,"resolution":{"self":"https://issues.apache.org/jira/rest/api/2/resolution/1","id":"1","description":"A fix for this issue is checked into the tree and tested.","name":"Fixed"},"customfield_12310220":"2006-04-26 04:55:54.0","customfield_12312322":null,"customfield_12312323":null,"customfield_12310420":"22380","customfield_12310222":"1_*:*_1_*:*_2751919000_*|*_6_*:*_1_*:*_0_*|*_5_*:*_1_*:*_684816000","customfield_12312320":null,"customfield_12312321":null,"customfield_12312120":null,"customfield_12312121":null,"resolutiondate":"2006-05-16T06:36:41.000+0000","workratio":-1,"customfield_12312328":null,"customfield_12312329":null,"customfield_12312326":null,"customfield_12310300":null,"customfield_12312327":null,"customfield_12312324":null,"customfield_12312325":null,"lastViewed":null,"watches":{"self":"https://issues.apache.org/jira/rest/api/2/issue/DERBY-1219/watchers","watchCount":0,"isWatching":false},"created":"2006-04-14T10:11:22.000+0000","priority":{"self":"https://issues.apache.org/jira/rest/api/2/priority/4","iconUrl":"https://issues.apache.org/jira/images/icons/priorities/minor.png","name":"Minor","id":"4"},"labels":[],"customfield_12312333":null,"customfield_12312334":null,"customfield_12310310":"10.0","customfield_12312331":null,"customfield_12312332":null,"timeestimate":null,"aggregatetimeoriginalestimate":null,"versions":[{"self":"https://issues.apache.org/jira/rest/api/2/version/11187","id":"11187","description":"","name":"10.2.1.6","archived":false,"released":true,"releaseDate":"2006-10-02"}],"customfield_12311120":null,"customfield_12312330":null,"issuelinks":[],"customfield_12312339":null,"assignee":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"customfield_12312337":null,"customfield_12312338":null,"updated":"2009-07-01T00:13:01.712+0000","customfield_12312335":null,"customfield_12311720":null,"customfield_12312336":null,"status":{"self":"https://issues.apache.org/jira/rest/api/2/status/6","description":"The issue is considered finished, the resolution is correct. Issues which are closed can be reopened.","iconUrl":"https://issues.apache.org/jira/images/icons/statuses/closed.png","name":"Closed","id":"6","statusCategory":{"self":"https://issues.apache.org/jira/rest/api/2/statuscategory/3","id":3,"key":"done","colorName":"green","name":"Complete"}},"components":[{"self":"https://issues.apache.org/jira/rest/api/2/component/11690","id":"11690","name":"Network Client"},{"self":"https://issues.apache.org/jira/rest/api/2/component/11410","id":"11410","name":"Network Server"}],"timeoriginalestimate":null,"description":"The tests checkDataSource.java and checkDataSource30.java \nhang intermittently especially with jdk 1.5.\n\nAttached is the test run output and traces when the server is started separately.\n\n1) Enable checkDataSource30.java by taking it out of functionTests/suites/DerbyNetClient.exclude.\n\n2) Run the test with client.\njava -Dij.exceptionTrace=true -Dkeepfiles=true -Dframework=DerbyNetClient org.apache.derbyTesting.functionTests.harness.RunTest jdbcapi/checkDataSource30.java\n\nAttachements:\n testfiles_after_hang.zip - Test directory.\n\n traces_on_hang.txt  - Server side traces obtained by starting the server separately before running the test.\n\nI wish I had time to work on this right now as I would really like to see this valuable test in the suite, but hopefully someone else will pick it up.\n\n\n","customfield_10010":null,"timetracking":{},"customfield_12312026":null,"customfield_12310200":null,"customfield_12312023":null,"customfield_12312024":null,"customfield_12312340":null,"aggregatetimeestimate":null,"customfield_12312022":null,"customfield_12312341":null,"customfield_12312220":null,"customfield_12310921":null,"customfield_12310920":"38558","summary":"jdbcapi/checkDataSource.java and jdbcapi/checkDataSource30.java hang intermittently with client","creator":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=kmarsden","name":"kmarsden","emailAddress":"kmarsdenderby at sbcglobal dot net","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Kathey Marsden","active":true},"subtasks":[],"customfield_12310090":null,"reporter":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=kmarsden","name":"kmarsden","emailAddress":"kmarsdenderby at sbcglobal dot net","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Kathey Marsden","active":true},"customfield_12310291":null,"customfield_12310290":null,"aggregateprogress":{"progress":0,"total":0},"environment":"More often on jdk 1.5 or jdk 1.6 but hangs on jdk 1.4.2 as well","customfield_12311020":null,"customfield_12310050":null,"duedate":null,"customfield_12310250":null,"progress":{"progress":0,"total":0},"comment":{"startAt":0,"maxResults":32,"total":32,"comments":[{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12374454","id":"12374454","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=kmarsden","name":"kmarsden","emailAddress":"kmarsdenderby at sbcglobal dot net","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Kathey Marsden","active":true},"body":"test output after hang.\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=kmarsden","name":"kmarsden","emailAddress":"kmarsdenderby at sbcglobal dot net","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Kathey Marsden","active":true},"created":"2006-04-14T10:12:37.000+0000","updated":"2006-04-14T10:12:37.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12374456","id":"12374456","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=kmarsden","name":"kmarsden","emailAddress":"kmarsdenderby at sbcglobal dot net","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Kathey Marsden","active":true},"body":"server side stack traces on hang.\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=kmarsden","name":"kmarsden","emailAddress":"kmarsdenderby at sbcglobal dot net","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Kathey Marsden","active":true},"created":"2006-04-14T10:15:27.000+0000","updated":"2006-04-14T10:15:27.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12376347","id":"12376347","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"body":"For DERBY-1148, I was trying to run checkDataSource test with client to repro the issue. I am getting the hang reported by Kathey almost all the time. So I looked into the hang and narrowed it down to this - The hang is occuring in ClientDataSource.getConnection method at this line: return ClientDriver.getFactory().newNetConnection((NetLogWriter) dncLogWriter, user, password, this, -1, false); \r\nIt looks like it is hanging in the method NetConnection.flowServerAttributesAndKeyExchange. I tried to capture server traces but running few times with trace on did not produce the hang for me. I'll be working with this test some more and will update if I get any info. If anyone else gets any clues as to what could be wrong, please post it.\r\n\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"created":"2006-04-26T04:55:54.000+0000","updated":"2006-04-26T04:55:54.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12377616","id":"12377616","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"body":"I talked to Kathey about this issue on IRC and she suggested trying this test outside the harness. I could repro it outside the harness too and I am attaching the stack traces from the client and the server (client_stack_trace_050306.txt, server_stack_trace_050306.txt) . From the trace and from the debugging I did with a few print statements, I think this is what is happening in the test:\r\n\r\n* For getting a new connection, client sends EXCSAT and ACCSEC to the server. Then, client waits for the response from the server for these commands. In all the test runs, I found the hang always happens in this request from the client sent as part of getConnection method. \r\n\r\n* At the server, it tries to read the data from the client and does not get any data in DDMReader.fill. So the server thinks the client has disconnected and ends the connection thread. \r\n\r\n* Client does not get back any response as server has ended the connection thread. Thus the client is blocked trying to read the input stream.\r\n\r\nI have not yet figured out the reason for this miscommunication between server and client. I would appreciate if someone can also go through the traces and confirm my analysis is correct.","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"created":"2006-05-04T02:41:33.000+0000","updated":"2006-05-04T02:41:33.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12377671","id":"12377671","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=bryanpendleton","name":"bryanpendleton","emailAddress":"bpendleton dot derby at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Bryan Pendleton","active":true},"body":"Hi Deepa,\r\n\r\nCan you get the DRDA trace files for the hang? I.e., set traceDirectory on the client and derby.drda.traceAll on the server, as described in the Protocol Tracing section of  http://wiki.apache.org/db-derby/ProtocolDebuggingTips","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=bryanpendleton","name":"bryanpendleton","emailAddress":"bpendleton dot derby at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Bryan Pendleton","active":true},"created":"2006-05-04T08:05:01.000+0000","updated":"2006-05-04T08:05:01.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12377814","id":"12377814","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"body":"Thanks Bryan. I am glad that you are looking at this issue. Hopefully we can narrow this down quickly with your DRDA expertise. I am attaching the client and server DRDA traces that I had captured previously when I ran the test inside the harness. You may find some traces which start with \"DERBY-1219\" which I had added for some debugging. Please ignore these. If needed, I can try to capture fresh traces and upload them.\r\n\r\nI had let the test run for a long time. At the end of the test I killed network server and this accounts for the last DisconnectException seen in client_trace.txt_sds_1. To me, it looked like the client was hanging after sending EXCSAT and ACCSEC whereas network server was idle. \r\n\r\n\r\n\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"created":"2006-05-04T21:49:20.000+0000","updated":"2006-05-04T21:49:20.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12378038","id":"12378038","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=bryanpendleton","name":"bryanpendleton","emailAddress":"bpendleton dot derby at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Bryan Pendleton","active":true},"body":"Hi Deepa, nothing obvious has jumped out at me yet, but I will keep looking. Three questions:\r\n\r\n1) You said you were able to reproduce this outside the harness; can you post a brief description of the steps, so that I can experiment with the code in my environment?\r\n2) Were you running with sane=true, or sane=false? If sane=false, can you try sane=true?\r\n3) You mentioned that the server does not get any data in DDMReader.fill, and thinks the client has disconnected. Can you expand on how you came to that conclusion? Were you able to see an IOException being thrown? Can you get a dump of that exception? After the exception has been thrown (i.e., during the hang), does netstat think that there is still an active TCP/IP connection between the client and the server?\r\n\r\nthanks, bryan\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=bryanpendleton","name":"bryanpendleton","emailAddress":"bpendleton dot derby at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Bryan Pendleton","active":true},"created":"2006-05-05T21:03:16.000+0000","updated":"2006-05-05T21:03:16.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12378079","id":"12378079","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"body":"Hi Bryan,\r\n\r\nHere are the answers to your questions:\r\n\r\n1) You said you were able to reproduce this outside the harness; can you post a brief description of the steps, so that I can experiment with the code in my environment?\r\n\r\nTo run the test jdbcapi/checkDataSource.java without using the test harness, I started network server and then used the following command to run the test:  \r\njava -Dderby.system.home=C:\\deepa\\Derby\\derby_testing\\nwserver -Dframework=DerbyNetClient -Dij.database=jdbc:derby://localhost:1527/wombat;create=true org.apache.derbyTesting.functionTests.tests.jdbcapi.checkDataSource\r\n\r\nThe test hangs intermittently and the place where the test hangs is in one of the getConnection methods. The hang location varies in different runs but is always in the getConnection method. I can repro it quite easily (~ 1 out of 5 runs hang) on my machine. I hope you are able to repro it too.\r\n\r\n2) Were you running with sane=true, or sane=false? If sane=false, can you try sane=true?\r\n\r\nI was running with sane=true. The debug output can be seen in the derby.log in the attached zip file (drda_traces_050206.zip)\r\n\r\n3) You mentioned that the server does not get any data in DDMReader.fill, and thinks the client has disconnected. Can you expand on how you came to that conclusion? Were you able to see an IOException being thrown? Can you get a dump of that exception? After the exception has been thrown (i.e., during the hang), does netstat think that there is still an active TCP/IP connection between the client and the server? \r\n\r\nI was looking at the server debug trace from a normal run (without hang) and when the test hangs. On comparing both and also looking at the client and server traces, I found the following from the trace files:\r\n* Client trace stops with send of EXCSAT and ACCSEC (client_trace.txt_sds_1)\r\n* At the server side, I can see an empty trace file is created but cannot see the EXCSAT and ACCSEC as the last entries in any trace file (Server10.trace, Server9.trace)\r\n* In the debug trace of server, the last trace is \"Ending connection thread\". I added few other traces in DDMReader.fill and DRDAConnThread.run  and found that server is actually reaching the end of the input stream (actualBytesRead == -1) without reading any data (totalBytesRead is 0) and hence calls \"agent.markCommunicationsFailure (\"DDMReader.fill()\", \"InputStream.read()\", \"insufficient data\", \"*\");\" The connection thread catches this disconnect exception and exits normally thinking the client has disconnected. The debug traces from a successful run show  that network server starts a new connection thread at this point.\r\n\r\nFrom the above, I thought of following possibilities:\r\n* server is not getting a set of data from the client. data is lost. this looks unlikely \r\n* server thread is reading a wrong stream (from an already closed connection) and thinking the client does not have more data and that it has disconnected. \r\n* session state associated with the server thread is wrong. Instead of starting new session, server is thinking it has an active session and tries to process commands for the active session\r\n\r\nI am just throwing in some ideas which come to me and could be totally off here. I plan to look at this some more and will post if I find something else.\r\n\r\nThere are no IOExceptions at server or client. The IO exception seen in the client trace file (client_trace.txt_sds_1) is because I killed network server after the test was hanging for a long time. This was just to reconfirm where the client is hanging. \r\n\r\nIn netstat output, I can see listening and established statuses for both client and server process during the hang. \r\n\r\nOuput of runtimeinfo command: \r\n--- Derby Network Server Runtime Information ---\r\n---------- Session Information ---------------\r\nSession # :10\r\n\r\n\r\n-------------------------------------------------------------\r\n# Connection Threads : 2\r\n# Active Sessions : 1\r\n# Waiting  Sessions : 0\r\n\r\nTotal Memory : 3694592  Free Memory : 2649768\r\n\r\nHope this helps.","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"created":"2006-05-06T00:34:24.000+0000","updated":"2006-05-06T00:34:24.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12378175","id":"12378175","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"body":"I forgot to mention that when I set the tracing for client, I was getting a NPE if I did not set all the ClientDataSource properties. Other option was to make a small change in client's LogWriter.java. I have opened DERBY-1298 for this. Just mentioning in case someone else hits the same problem when tracing the client.","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"created":"2006-05-06T07:13:32.000+0000","updated":"2006-05-06T07:13:32.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12378228","id":"12378228","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=bryanpendleton","name":"bryanpendleton","emailAddress":"bpendleton dot derby at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Bryan Pendleton","active":true},"body":"Hi Deepa. Thank you for the good notes. As you say:\r\n\r\n> The test hangs intermittently and the place where the test hangs is in one of the getConnection\r\n>   methods. The hang location varies in different runs but is always in the getConnection method. \r\n>   I can repro it quite easily\r\n\r\nI appear to be able to reproduce it quite easily, too. So that is good.\r\n\r\nI think that an interesting aspect of this test is that it creates and tears down a lot of connections\r\nin very rapid succession, and I am wondering whether there is a race condition somewhere\r\nthat is causing the code to lose track of which connection is which.\r\n\r\nI think that your analysis of the (actualBytesRead == -1) case is good, but is perhaps a red\r\nherring. I believe that this is the normal way that the server cleans up when a client\r\nconnection disconnects and goes away. So I don't think this is *directly* the place where things\r\nare going wrong; it's just evidence that the server is seeing a lot of connections come and go.\r\n\r\nIt seems like, at least in the cases I've seen so far, the bottom line is this:\r\n - the client believes it's initiated a new connection with the server\r\n - the server, however, has no record of that connection, and believes it's cleaned up all\r\n   its connections and is idle\r\n\r\nSo I think something is going wrong in the connection management code on one side or\r\nthe other.\r\n\r\nThanks for the great test case and set of notes; I'll continue to study this one some more and\r\nlet you know if I figure anything out.\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=bryanpendleton","name":"bryanpendleton","emailAddress":"bpendleton dot derby at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Bryan Pendleton","active":true},"created":"2006-05-06T23:03:21.000+0000","updated":"2006-05-06T23:03:21.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12378300","id":"12378300","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=bryanpendleton","name":"bryanpendleton","emailAddress":"bpendleton dot derby at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Bryan Pendleton","active":true},"body":"Attached are \"skipThreads.diff\" and \"interrupt.diff\", but before\r\nreading the diffs, please read these notes.\r\n\r\nI think I understand what is causing the hangs, and I can even make\r\nthe hangs go away. However, I don't think I yet understand how to\r\nreally *fix* the problem, so I'm sure we'll want to talk about this\r\nfor a while, to see if some of the reviewers can come up with a\r\nproper solution or at least some techniques to pursue.\r\n\r\nHere's what I see, and what I think it means:\r\n\r\n1) One, or maybe several, times in the test, checkDataSource causes a \r\nshutdown of the server. It has several different variants on the shutdown\r\nprocessing, but at least one of them causes the server to go through\r\nNetworkServerControlImpl.startNetworkServer() to perform a server restart.\r\n\r\n2) During the server restart processing, the Network Server restart \r\ncode iterates through all the DRDAConnThread instances and closes them. \r\nThis close() call is supposed to cause the DRDAConnThread to terminate itself.\r\n\r\n3) However, all the close() call actually does is mark the thread's\r\n\"close\" variable as true, and depending on when the thread checks that\r\nvariable, it may or may not immediately exit. In my test runs, it is\r\noften the case that at least one of the DRDAConnThread instances is, at this\r\npoint, sitting blocked in NetworkServerControlImpl.getNextSession().\r\nCalling close() on this thread marks it as closed, but doesn't cause\r\nit to exit the getNextSession() wait.\r\n\r\n4) A little bit later, the test program makes some new connections\r\nto the server, and one of those connections is given to the thread\r\nwhich was blocked in the getNextSession() call. The thread picks\r\nup the session and returns to the DRDAConnThread.run() main loop.\r\n\r\n5) At this point, the thread notices that it has been closed, and it\r\nexits, without sending any response back to the client, and without\r\nclosing the connection to the client. This causes the hang.\r\n\r\nBecause this problem involves multi-threading, and thread scheduling,\r\nthere is a bunch of non-determinate behavior, which I believe is why\r\nothers have been experiencing varied results during their tests. The\r\nbehavior of the threads is definitely unpredictable for me.\r\n\r\nThere are several aspects to this scenario that puzzle me, but let\r\nme describe what I've been experimenting with as a patch. I've changed\r\nthe NetworkServerControlImpl restart logic so that, instead of\r\nclosing the DRDAConnThreads, it just leaves the threads alone.\r\nThis change is in \"skipThreads.diff\", and it seems to make the hangs\r\ngo away. \r\n\r\nThe \"skipThreads.diff\" diff also contains some hacks to the test so\r\nthat I could run it multiple times in a row outside of the harness\r\nwithout destroying and re-creating the database each time.\r\nThose changes don't really belong with this diff, but I didn't bother\r\nto edit them out.\r\n\r\nI also experimented with a change which tried to close the threads,\r\nbut also, after closing, interrupts the threads, which\r\ncaused them to be blown out of the getNextSession loop and back to\r\nthe main run() loop, at which point the threads shut themselves down,\r\nwhich seems like the right behavior for Network Server restart. \r\n\r\nI was hoping that this was the \"right\" fix, but unfortunately this\r\nchange fixed *some*, but not *all*, of the hangs, which was too bad.\r\nAnd I'm nervous about adding the call to Thread.interrupt(), which is\r\nan extremely powerful call and not to be used lightly. For reviewers\r\nwho want to experiment with this change and see how it works for them,\r\nI've also attached \"interrupt.diff\"\r\n\r\nI'm still disturbed by the fact that when the main run() method\r\nin DRDAConnThread noticed that it was closed, it just exited without\r\napparently sending any response back to the server or closing the\r\nsocket.\r\n\r\nAnd, although my change makes the hangs go away, it does not make the\r\ncheckDataSource and checkDataSource30 tests pass. Instead, they run\r\nto completion, and get a bunch of diffs, and I'm not sure whether my\r\nchanges caused these diffs or not.\r\n\r\nBut at this point, before I work on this much more, I'd like to get\r\nsome feedback from the reviewers about the analysis up to this point,\r\nand the effects of this patch in their environment:\r\n\r\n - does this patch cause the hangs to disappear for you?\r\n - if so, do the checkDataSource and checkDataSource30 tests pass for you?\r\n - if they fail, do the failures make sense to you?\r\n - what should we be doing with the background connection threads during\r\n   a Network Server restart?\r\n\r\nThanks!","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=bryanpendleton","name":"bryanpendleton","emailAddress":"bpendleton dot derby at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Bryan Pendleton","active":true},"created":"2006-05-07T22:31:54.000+0000","updated":"2006-05-07T22:31:54.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12378481","id":"12378481","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"body":"I ran the standalone repro with both your patches. I was not able to reproduce the hang on my machine after 25 runs with each patch.\r\n\r\nThanks for the detailed explanation of the problem. I think I was reading some overlapping traces and was misled. After reading your description, I added more traces and confirmed that the run method breaks out because it finds that the thread has been marked closed. As you said, it seems not quite okay to just break out without informing the client. Maybe, it is expected that this code can reached only after client is already informed after an exception or at end of a session. A quick look at places where close method is called seems to indicate this. \r\n\r\nBoth your solutions seem to work on my machine but I am not very clear about them. About solution 1 (skipThreads.diff) , it does not seem quite right to remove the cleanup code (code to close existing threads and clear threadList). If it is okay to reuse the same threads after a restart, then it should be okay to remove this cleanup code. Even then, I think the real cause of this problem could be somewhere else. I have not looked at solution 2 as you mentioned it does not work in all cases.\r\n\r\nJust sharing another thought which occurred to me after reading your descriptions. It looks like the new thread which was opened to serve the new connection gets added to threadList before the list is going to be cleaned up during restart. This causes the new thread to be closed unexpectedly when the server is restarted. I do not want to lead you down the wrong path. So I am looking at how/where threadList is accessed and if there could me some missing synchronizations here. I will post if I find something.","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"created":"2006-05-09T01:01:32.000+0000","updated":"2006-05-09T01:01:32.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12378734","id":"12378734","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=bryanpendleton","name":"bryanpendleton","emailAddress":"bpendleton dot derby at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Bryan Pendleton","active":true},"body":"One strategy that occurred to me is that, rather than clearing and re-using the RunQueue and the ThreadList, \r\nwe could have the restart processing set up a new RunQueue and ThreadList, and that way make a clean\r\ndistinction between old sessions and threads, versus new sessions and threads. That is, try to avoid the\r\nproblem of \"a new session comes in, but gets handled by an old thread which is being closed\", by teaching\r\nthe code how to tell the difference between old and new threads and sessions.\r\n\r\nSome sort of restart algorithm like:\r\n\r\n  oldRunQueue = RunQueue\r\n  oldThreadList = ThreadList\r\n  create new RunQueue and new ThreadList\r\n  go through the old RunQueue and thread list and close and clean them up\r\n\r\nThe idea being that, once we reach a certain point in Network Server restart processing, all new\r\nthreads go onto the new thread list, and all new sessions go onto the new run queue, and it\r\nshould be easier to tell the difference between old threads & sessions, which have been closed\r\nand should now die, and new threads and sessions, which are allowed to start doing new work.\r\n\r\nA similar, but not identical, idea would be to establish some sort of \"generation\" counter, which\r\nis incremented by one each time the Network Server restarts within a process, and instead\r\nof using the \"close\" API to shut down sessions and threads, use the idea of generations, so\r\nthat a session or thread which is in an older generation is treated as dead and gets cleaned up,\r\nwhile new generation threads can start processing new generation sessions, and if a thread\r\nand a session ever found that their generations didn't match, they would know that something\r\nhorrible had occurred and they could trigger a sanity check or assertion.\r\n\r\nAnyway, just some thoughts to keep the discussion going. \r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=bryanpendleton","name":"bryanpendleton","emailAddress":"bpendleton dot derby at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Bryan Pendleton","active":true},"created":"2006-05-10T03:33:23.000+0000","updated":"2006-05-10T03:33:23.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12378767","id":"12378767","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"body":"I am thinking along slightly different lines. The restart processing is doing some cleanup and reloading the driver. Should'nt we wait for the reload to happen before we start working on the new session which comes in? In that case, we will need some logic to check if a restart is in progress before we add a new thread/session to the lists.\r\n\r\nOr do we expect the driver to be reloaded by the time the new thread/session gets ready to do the real work? \r\n\r\n\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"created":"2006-05-10T05:31:38.000+0000","updated":"2006-05-10T05:31:38.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12378772","id":"12378772","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=bryanpendleton","name":"bryanpendleton","emailAddress":"bpendleton dot derby at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Bryan Pendleton","active":true},"body":"I think that waiting for the reload to complete is fine, but we'll need to figure out a way to have a\r\npositive confirmation that the threads have received their notification and shut themselves down,\r\nsince, from a certain point of view, the essence of this bug is that calling DRDAConnThread.close()\r\nsimply asks a thread to shut itself down, but does not wait for that shutdown to actually occur.\r\n\r\nSuch an algorithm would look something like:\r\n  - call close on each thread\r\n  - call interrupt on each thread\r\n  - while the threadlist is not empty\r\n        wait for a while\r\n\r\nI'm always nervous about algorithms like this because of the need for an open-ended wait:\r\n - how do we know how long to wait?\r\n - what happens if we've waited a long time and the threads still haven't shut themselves down?\r\n\r\nThe algorithms that I was proposing in my previous comment basically replace this\r\nopen-ended wait logic with a potential resource leak in the scenario where the old\r\nthreads for some reason don't respond to the request to shut themselves down.\r\nThat is, either we accept the risk of leaking away old threads, or we accept the risk\r\nof waiting indefinitely for old threads, (or, I suppose, we figure out some way to be\r\nVERY confident that we can shut the old threads down in a reasonable amount of time).\r\n\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=bryanpendleton","name":"bryanpendleton","emailAddress":"bpendleton dot derby at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Bryan Pendleton","active":true},"created":"2006-05-10T06:00:29.000+0000","updated":"2006-05-10T06:00:29.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12378794","id":"12378794","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"body":"Hi Bryan, \r\nWe seem to be thinking of the problem differently. I am just thinking aloud to explain how I understood the scenario:\r\n\r\nI think the scenario is: \r\n\r\nOne of the threads have asked to restart the server and this \"restart\" includes:\r\n1. a) close all sessions in runQueue b) clear runQueue\r\n2. a) close all threads in threadList b) clear threadList \r\n3. reload the embedded driver\r\n\r\n(Note: 1a and 2a are not single steps)\r\n\r\nThis restart is in progress and we get a \"new connection\" from the client. We can do two things:\r\n1. queue it by adding to the runQueue list\r\nOR\r\n2. a) create a new thread for the connection b) add it to threadList c) start the new connection thread\r\n\r\nIn addition to this, we can also have other connection threads which are running.\r\n\r\nSome cases I can think of which can cause hang are if we have actions overlapped as follows:\r\ncase1: restart_1a, restart_1b, new_conn_2a, new_conn_2b, restart_2a, new_conn_2c\r\ncase2: new_conn_1, restart 1a, restart 1b ...\r\n\r\nIn case 1, after we have added a new connection thread to the threadList, threadList is cleared by restart thread and all threads including the new thread are marked closed. We then start the new thread (new_conn_2c) and it exits because it finds it was marked closed.\r\n\r\nIn case 2, we queue up the connection request, runQueue is cleared during restart. So we miss the request.\r\n\r\nTo me, the problem seems to be that we access the lists runQueue and threadList when a restart is in progress. \r\n\r\nIn case of thread shutdown, I think the way it currently works is that we mark the thread as closed (set close = true). And next time, this thread is used, it checks if it was closed and if so it ends itself without doing anything. As the thread will not do anything else once it is marked closed, I am thinking it is okay to consider it as closed and not wait for it to actually terminate.  To me, the culprits seem to be the lists maintained by the server.\r\n\r\nAny thoughts/comments?","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"created":"2006-05-10T08:07:56.000+0000","updated":"2006-05-10T08:07:56.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12378821","id":"12378821","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=bryanpendleton","name":"bryanpendleton","emailAddress":"bpendleton dot derby at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Bryan Pendleton","active":true},"body":"Hi Deepa,\r\nI think that your scenarios are excellent, and they definitely demonstrate the problems in this area. \r\nI think that the fun thing about a bug like this, is that there are many possible scenarios. :)\r\n\r\nThe one I was concentrating on is a bit different from yours, so let me try to diagram it as follows:\r\n\r\n1) Some thread is idling, blocked in NetworkServerControlImpl.getNextSession() as called by\r\n   DRDAConnThread.run(). I think this is the standard place for a thread to block when it is idle.\r\n\r\n2) Server restart occurs, and runs straight through to completion. This results in calling close()\r\n   on the thread from point (1), and also removing that thread from the ThreadList. *But the thread\r\n   does not terminate.*\r\n\r\n3) Some time later, some new connections start coming in. The first new connection, as you\r\n   point out, will create a new thread to handle the session. The next new conection, however, will\r\n   find that a thread already exists, and so it will simply put the session onto the RunQueue list.\r\n\r\n4) The original thread then wakes up, grabs the session, notices that the thread has been\r\n   closed, and exits.\r\n\r\nThe point I'm trying to make here is that no overlapping of actions is required, and the connection \r\ndoes not have to arrive during the restart. \r\n\r\nIt seems to me that, once a restart happens while 1 or more threads happen to be sitting idle,\r\nblocked in their getNextSession() calls, then those threads are \"poisoned\", and there is\r\nnow a ticking time bomb in the server. At some point in the future, a session will get added\r\nto the RunQueue, and one of these \"poisoned\" (closed) threads will grab the session, and\r\nwill then terminate prematurely without processing the session.\r\n\r\nSo the only place where I differ with your analysis, I believe, is that I think it is *not* okay to\r\nleave these threads out there, marked as closed, because at some point in the future the\r\nthreads will grab sessions off the run queue and fail to process them.\r\n\r\nSo I think one crucial thing to ensure is that, once a thread is marked as closed, it will no \r\nlonger pick up a new session to process.\r\n\r\nWith that in mind, I've experimented with yet another patch, called \"no-sessions-for-closed-threads.diff\",\r\nwhich attempts to prevent threads marked as closed from fetching sessions to process.\r\n\r\nIt seems to resolve the hang for me, but I haven't exhaustively tested it. Still, I thought it\r\nshowed enough promise to attach for you to examine.\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=bryanpendleton","name":"bryanpendleton","emailAddress":"bpendleton dot derby at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Bryan Pendleton","active":true},"created":"2006-05-10T11:30:51.000+0000","updated":"2006-05-10T11:30:51.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12378938","id":"12378938","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"body":"Hi Bryan,\r\nI applied your new patch 'no-sessions-for-closed-threads.diff' and ran the checkDataSource repro. However, I still get the hang intermittently on my machine. Though the hang has not gone away, I think the scenario you explain is a probable one. And as you said, the bug is quite interesting as it can have many possible scenarios.\r\n\r\n \r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"created":"2006-05-10T22:34:25.000+0000","updated":"2006-05-10T22:34:25.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12378954","id":"12378954","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"body":"Posting my observation with the patch 'no-sessions-for-closed-threads.diff': I think this patch solves the problem of \"poisoned\" threads partially. It makes sure that a thread which has been marked closed does not get a session to work on. However, the session which came in is still hanging because there are no new threads which can pick it up. When the session came in, server had queued it so that a free thread picks it up. But the thread which picked it up was a \"poisoned\" (closed) thread. So the session is still waiting for another thread. \r\n\r\nTo test that the hang will go away when a new thread gets created, I opened another connection using ij. This made the hang go away and the repro ran to the end. When the connection from ij came in, the newly created thread was able to work on the waiting session and resolve the hang.\r\n\r\nBryan, I am looking at your trial patch \"interrupt.diff\" which looks like a good solution to me too. This patch made the hang go away on my machine. But you had said that it did not resolve all hangs for you. Can you please post your observations with this patch?","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"created":"2006-05-11T00:32:53.000+0000","updated":"2006-05-11T00:32:53.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12378971","id":"12378971","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=bryanpendleton","name":"bryanpendleton","emailAddress":"bpendleton dot derby at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Bryan Pendleton","active":true},"body":"I will gladly work on interrupt.diff some more, and try to provide more testing details.\r\nBut I won't get to it until at least this weekend, sorry. Thanks very much for continuing\r\nto work on this with me; I think we're making some real progress here!\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=bryanpendleton","name":"bryanpendleton","emailAddress":"bpendleton dot derby at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Bryan Pendleton","active":true},"created":"2006-05-11T03:19:56.000+0000","updated":"2006-05-11T03:19:56.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12383400","id":"12383400","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=bryanpendleton","name":"bryanpendleton","emailAddress":"bpendleton dot derby at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Bryan Pendleton","active":true},"body":"I've been trying to run this test in one of my alternate environments, and I'd like to be able to run the server on a port other than 1527. Unfortunately, when I change the value of ij.database so that it specifies a different port number, the test runs partway, then fails with:\r\n\r\njava.sql.SQLException: java.security.PrivilegedActionException : Error connecting to server localhost on port 1,527 with message Connection refused.\r\n        at org.apache.derby.client.am.SQLExceptionFactory.getSQLException(SQLExceptionFactory.java:45)\r\n        at org.apache.derby.client.am.SqlException.getSQLException(SqlException.java:342)\r\n        at org.apache.derby.jdbc.ClientDataSource.getConnection(ClientDataSource.java:191)\r\n        at org.apache.derby.jdbc.ClientDataSource.getConnection(ClientDataSource.java:162)\r\n        at org.apache.derbyTesting.functionTests.tests.jdbcapi.checkDataSource.runTest(checkDataSource.java:190)\r\n        at org.apache.derbyTesting.functionTests.tests.jdbcapi.checkDataSource.main(checkDataSource.java:132)\r\n\r\nLine 190 of checkDataSource.java is the last line of the following snippet:\r\n\r\n                DataSource dscs = TestUtil.getDataSource(attrs);\r\n\r\n                if (testConnectionToString)\r\n                                checkToString(dscs);\r\n\r\n                DataSource ds = dscs;\r\n\r\n                checkConnection(\"DataSource\", ds.getConnection());\r\n\r\nIs it possible to configure this test to run on a port other than 1527?\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=bryanpendleton","name":"bryanpendleton","emailAddress":"bpendleton dot derby at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Bryan Pendleton","active":true},"created":"2006-05-14T00:29:08.000+0000","updated":"2006-05-14T00:29:08.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12383402","id":"12383402","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=bryanpendleton","name":"bryanpendleton","emailAddress":"bpendleton dot derby at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Bryan Pendleton","active":true},"body":"I think the interrupt.diff patch is very close to being good, but I think there are some \r\nsynchronization issues to resolve. The interrupt.diff patch is very reliable on my\r\nWindows environment; that is, I cannot reproduce the hang with that diff in place.\r\n\r\nIn my Linux environment, however, the test still hangs with that patch. However, if\r\nI introduce even very small timing changes into the test; for example if I put a\r\nprintln statement into a critical section of the getNextSession() logic, then the\r\nhang disappears.\r\n\r\nSo I think that the interrupt.diff patch can perhaps be combined with some\r\nadditional synchronization as Deepa suggested earlier, to solve some of the\r\nraces between the Network Server restart, the DRDAConnThreads calling\r\ngetNextSession, and the ClientThread logic involving creating a new session\r\nand conditionally creating a new thread.\r\n\r\nI'll continue to fiddle with the configuration that I have in which I can still provoke\r\na hang, and see what I can do to understand it better.\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=bryanpendleton","name":"bryanpendleton","emailAddress":"bpendleton dot derby at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Bryan Pendleton","active":true},"created":"2006-05-14T00:47:53.000+0000","updated":"2006-05-14T00:47:53.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12402243","id":"12402243","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=bryanpendleton","name":"bryanpendleton","emailAddress":"bpendleton dot derby at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Bryan Pendleton","active":true},"body":"Well, I ran a lot of experiments, but I don't have anything conclusive to offer. I believe that:\r\n - the issue is intricately linked to the handling of threads and sessions during restart\r\n - the hangs are because closed threads are given sessions to run, but then abandon\r\n   those sessions without running them when they discover they've been closed\r\n - the interrupt() call makes things better, but does not totally fix the problem. I can still\r\n   reproduce the problem on my RedHat Linux/ Sun JDK 1.4.2 environment.\r\n - but my various attempts to try to improve on the interrupt() fix with additional changes,\r\n   only made things worse.\r\n\r\nUnfortunately, I'm not going to have much more time to work on this right away, so I'm\r\nunassigning myself in the hopes that somebody else can work on it.\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=bryanpendleton","name":"bryanpendleton","emailAddress":"bpendleton dot derby at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Bryan Pendleton","active":true},"created":"2006-05-15T05:58:05.000+0000","updated":"2006-05-15T05:58:05.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12402353","id":"12402353","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"body":"Thanks Bryan for helping to identify the cause of the hang. \r\n\r\nSince we now know what is causing the intermittent hang, I think we can break the issue as follows: \r\n\r\n1) Open a separate code bug against network server component to solve the real hang problem and link to the details in this test issue.\r\n\r\n2) Enable the test to run with client by removing the code which shutdowns the system when we run the test using client framework. This is the code which causes the intermediate hang and it seems this code is not required to check the scenarios meant to be covered by the test.  I disabled the following shutdown code using a boolean and ran the checkDataSource tests with the client and did not get the hang so far:\r\n\r\n          // DERBY -???? - link to new bug opened for 1) above\r\n          private static boolean hangAfterSystemShutdown = TestUtil.isDerbyNetClientFramework();\r\n                        .\r\n                        .\r\n                        .\r\n          // Shutdown the system only if we are not running in client framework\r\n          if(! hangAfterSystemShutdown) {\r\n\t\t\ttry {\r\n\t\t\t\tTestUtil.getConnection(\"\",\"shutdown=true\");\r\n\t\t\t} catch (SQLException sqle) {\r\n\t\t\t\tJDBCDisplayUtil.ShowSQLException(System.out, sqle);\r\n\t\t\t}\r\n\t\t}\r\n\r\nThe test has some existing diffs because of some intermediate check-ins which changed the messages/behaviour. I will look at these diffs and submit a patch to enable the checkDataSource tests to run with client. I think it would be good to get the test running with client so that we can keep the masters up to date and also be able to catch any regressions in this area.\r\n\r\nPlease provide any comments/feedback.","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"created":"2006-05-16T00:04:55.000+0000","updated":"2006-05-16T00:04:55.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12402400","id":"12402400","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"body":"The two categories of diffs seen  when running the checkDataSource test with client are: \r\n\r\n< DriverManager  <closedstmt>.execute() null - Invalid operation: statement closed\r\n11a11\r\n> DriverManager  <closedstmt>.execute() 08003 - No current connection\r\n\r\nAND\r\n\r\n< autocommitxastart expected 'ResultSet' already closed.\r\n460a459\r\n> autocommitxastart expected ResultSet not open.  Verify that autocommit is OFF.\r\n\r\nThese diffs are message changes done as part of DERBY-843/DERBY-842. However, in the case of first diff, I see a difference in the behaviour of client and embedded driver.\r\n\r\nIn case of client driver, Statement.execute method first checks if the connection associated with the statement is closed before it checks if the statement itself was closed or not. In case of embedded driver, the check for statement closed is done before the check for connection closed. I would like to clarify if the order of the checks matter and if it does which is the correct behaviour.","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"created":"2006-05-16T02:56:56.000+0000","updated":"2006-05-16T02:56:56.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12402409","id":"12402409","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=kmarsden","name":"kmarsden","emailAddress":"kmarsdenderby at sbcglobal dot net","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Kathey Marsden","active":true},"body":"Deepa said ...\r\n>(The shutdown)  is the code which causes the intermediate hang and it seems this code is not required >to check the scenarios meant to be covered by the test.  \r\n\r\nI agree that the conditional  is a good solution for getting the test running with client, but would like to clarify that I think the shutdown is a valuable part of the test because we verify that the global transaction state is valid even aver the database has been shutdown and restarted.  Once the hang has been resolved it would be good to reenable this part of the test for client.\r\n\r\n\r\n\r\n\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=kmarsden","name":"kmarsden","emailAddress":"kmarsdenderby at sbcglobal dot net","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Kathey Marsden","active":true},"created":"2006-05-16T03:17:45.000+0000","updated":"2006-05-16T03:17:45.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12402429","id":"12402429","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=kmarsden","name":"kmarsden","emailAddress":"kmarsdenderby at sbcglobal dot net","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Kathey Marsden","active":true},"body":"Deepa asked about these diffs:\r\n>DriverManager <closedstmt>.execute() null - Invalid operation: statement closed\r\n11a11\r\n> DriverManager <closedstmt>.execute() 08003 - No current connection\r\n\r\nAND\r\n\r\n< autocommitxastart expected 'ResultSet' already closed.\r\n460a459\r\n> autocommitxastart expected ResultSet not open. Verify that autocommit is OFF.\r\n\r\nrelated to the DERBY-843 change.  I posted a question about this to DERBY-843, but I think it makes sense to go ahead and update the master and get the test running while that issue is being resolved.\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=kmarsden","name":"kmarsden","emailAddress":"kmarsdenderby at sbcglobal dot net","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Kathey Marsden","active":true},"created":"2006-05-16T04:51:40.000+0000","updated":"2006-05-16T04:51:40.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12402442","id":"12402442","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"body":"Attaching a patch 'derby-1219-enable-tests.diff' which enables the tests jdbcapi/checkDataSource.java and jdbcapi/checkDataSource30.java to run with client.\r\n\r\nPatch adds a condition to exclude a system shutdown when running with client framework. This condition has to be removed once the hang (DERBY-1326) is resolved. Thanks Kathey for mentioning the use of shutdown in this test.\r\n\r\nWith this patch, I ran the 2 tests using embedded and client driver with Sun jdk1.4.2 on Windows XP. I ran the tests multiple times with client driver to check that the hang does not occur. I would appreciate if someone can take a look at this patch.\r\n\r\n\r\n\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"created":"2006-05-16T05:44:09.000+0000","updated":"2006-05-16T05:44:09.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12402450","id":"12402450","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=bryanpendleton","name":"bryanpendleton","emailAddress":"bpendleton dot derby at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Bryan Pendleton","active":true},"body":"I ran both tests multiple times on my Linux JDK 1.4.2 environment with no hangs, and tests passed\r\neach time. This particular machine was very reliable in reproducing the hang before, so I think that's\r\na good result, and it confirms your result.\r\n\r\nI would be glad to commit this patch. Do you think there are any other details we should check first?\r\n\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=bryanpendleton","name":"bryanpendleton","emailAddress":"bpendleton dot derby at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Bryan Pendleton","active":true},"created":"2006-05-16T06:08:48.000+0000","updated":"2006-05-16T06:08:48.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12402454","id":"12402454","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"body":"Were you asking if we need to run more tests? I think this patch changes only tests and I have run the tests with embedded and client driver. The tests are excluded with jcc driver. We also confirmed that the hang does not occur with the changed test.\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=deepa","name":"deepa","emailAddress":"dremesh at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Deepa Remesh","active":true},"created":"2006-05-16T06:26:00.000+0000","updated":"2006-05-16T06:26:00.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12402458","id":"12402458","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=bryanpendleton","name":"bryanpendleton","emailAddress":"bpendleton dot derby at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Bryan Pendleton","active":true},"body":"I committed patch derby-1219-enable-tests.diff to subversion as revision 406776:\r\nhttp://svn.apache.org/viewcvs?rev=406776&view=rev\r\n\r\nIt will be nice to have this test running regularly in the client framework.\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=bryanpendleton","name":"bryanpendleton","emailAddress":"bpendleton dot derby at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Bryan Pendleton","active":true},"created":"2006-05-16T06:36:41.000+0000","updated":"2006-05-16T06:36:41.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12332120/comment/12413008","id":"12413008","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=kmarsden","name":"kmarsden","emailAddress":"kmarsdenderby at sbcglobal dot net","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Kathey Marsden","active":true},"body":"Cannot reproduce hang. Thanks Deepa","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=kmarsden","name":"kmarsden","emailAddress":"kmarsdenderby at sbcglobal dot net","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Kathey Marsden","active":true},"created":"2006-05-24T04:50:17.000+0000","updated":"2006-05-24T04:50:17.000+0000"}]},"votes":{"self":"https://issues.apache.org/jira/rest/api/2/issue/DERBY-1219/votes","votes":0,"hasVoted":false},"worklog":{"startAt":0,"maxResults":20,"total":0,"worklogs":[]},"customfield_12311820":"0|i06ypr:"}}